crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -71,101 +75,182 @@ impl CoinGecko {
-                        self.send_request_price_in_eth(token).await?,
-                    )
-                    .map_err(|e| {
-                        PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse requested token in ETH to rust decimal: {e:?}"
-                        ))
-                    })?;
-                    let native_price_token_eth = rust_decimal::Decimal::try_from(
-                        self.send_request_price_in_eth(native_price_token).await?,
-                    )
-                    .map_err(|e| {
-                        PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse native price token in ETH to rust decimal: {e:?}"
-                        ))
-                    })?;
-                    let token_in_native_price =
-                        token_eth.checked_div(native_price_token_eth).ok_or(
-                            PriceEstimationError::EstimatorInternal(anyhow!("division by zero")),
-                        )?;
-                    token_in_native_price
-                        .to_f64()
-                        .ok_or(PriceEstimationError::EstimatorInternal(anyhow!(
-                            "failed to parse result to f64"
-                        )))
+                    if !requested_tokens.contains(&native_price_token) {
+                        tokens.push(&native_price_token);
+                    }
+
+                    self.bulk_fetch_denominated_in_token(tokens, native_price_token)
+                        .await
                 }
             }
         }
         .boxed()
     }
+
+    fn max_batch_size(&self) -> usize {
+        /// maximum number of price the coingecko API returns in a single batch
+        const MAX_BATCH_SIZE: usize = 20;
+
+        match self.quote_token {
+            QuoteToken::Eth => MAX_BATCH_SIZE,
+            // when fetching price denominated in a custom token we need to
+            // fetch the price for that token in addition to the requested
+            // tokens so we reserve 1 spot in the batch
+            QuoteToken::Other(_) => MAX_BATCH_SIZE - 1,
+        }
+    }
+}
+
+impl NativePriceEstimating for CoinGecko {
+    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let prices = self.fetch_native_prices(HashSet::from([token])).await?;
+            prices
+                .get(&token)
+                .ok_or(PriceEstimationError::NoLiquidity)?
+                .clone()
+        }
+        .boxed()
+    }
+}
+
+impl From<anyhow::Error> for PriceEstimationError {
+    fn from(err: anyhow::Error) -> PriceEstimationError {
+        PriceEstimationError::EstimatorInternal(err)
+    }
 }
 
 mod observe {
     use url::Url;
 
     /// Observe a request to be sent to CoinGecko
-    pub fn coingecko_request(endpoint: &Url) {
+    pub(super) fn coingecko_request(endpoint: &Url) {
```

Comment:
btw, fyi, this is redundant (no need to change, out of curiosity), `observe` is only accessible within the `coingecko.rs` file anyway.

Reply:
Yeah, I'm just trying to be conscious about the visibility of items to help the compiler find dead code and avoid leaking things into the public interface of a crate. Once things are there the compiler can't help us find dead code anymore. :/

Reply:
That's true, you are totally right